### PR TITLE
Make functional tests run faster

### PR DESCRIPTION
**Pull request by pmeinhardt**
#### Update `SSHKit` and `SSHKit.SCP` functional tests to be `async: true`

These tests are currently not run in concurrently with other tests. Both already use temp directories with randomly generated names for testing everything which touches the local system (downloads from a container).

However, to fully isolate them, we also need to avoid things like `File.cd!` which changes the working directory of all (test) processes and can cause parallel tests to fail.

Once they are made fully isolated, there is no need to set `async` to `false` anymore.

✅

Reported times for `mix test --only functional` showed a reduction from 35 s to 25 s on my machine, from 55 s to 39 s on Travis CI, when using `async: true`. Roughly a 28-30% improvement

#### 🚧 Question: Can we further isolate functional test processes?

The current assumption in our functional tests is that they take care to create temp directories whenever they wish to touch the local filesystem and that they clean up after themselves. We assume tests and SSHKit operations do not do anything completely crazy – which seems quite reasonable and is also done in the [Elixir test suite](https://github.com/elixir-lang/elixir/blob/8f0fe3d23f11e9d9461e3bcf6dfab72eb28526af/lib/elixir/test/elixir/file_test.exs#L13-L17).

However, maybe we can make functional tests even safer, isolating the local filesystem?
Maybe we can isolate functional tests so they're safe to use `File.cd` and the likes?

Brainstorming: run in containers, chroot, memory filesystem, filesystem overlay, mock `Elixir.File`…

❔ 

### Motivation and Context

It takes about 35 seconds to run our functional tests on the current `master`. Ideally they should run faster for a more pleasant development experience and quicker feedback.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes (with unit and/or functional tests).
- [ ] I have added a note to CHANGELOG.md if necessary (in the `## master` section).

### CODE REVIEW

**Comment by pmeinhardt** [comment]
I have not found a satisfying answer to the second question, but I am okay with that for now (I you have any ideas, shoot). Plus I have an image of @tessi in the back of my head who is whispering – with a ghost like voice: "Smaller PRs, smaller PRs…" 👻 

So, here ya go 🚶

File: test/sshkit_functional_test.exs [hunk 1]
```diff
@@ -184,59 +184,58 @@ defmodule SSHKitFunctionalTest do
   end
 
   describe "download/3" do
+    setup do
+      tmpdir = create_local_tmp_path()
+
+      :ok = File.mkdir!(tmpdir)
+      on_exit fn -> File.rm_rf(tmpdir) end
```
Comment (klappradla) [inline test/sshkit_functional_test.exs]: Clean solution, like that 🗑

**Comment by pmeinhardt** [comment]
@klappradla, thanks a lot for the review :metal: 

As for the open question: My idea was protecting the developer machine by preventing modification of the local filesystem rather than isolating tests from other tests.

The tests now run in parallel because we don't mutate shared state (no files used by multiple test processes and no changing of the system process's current working directory), so that problem is solved for now :white_check_mark:  However, an error in a test case could still remove files on your local fs or create them in places where you don't want them… :bomb: 

I was therefore :thinking: whether we should run the (functional) test suite *as a whole* within some kind of isolated environment.

It would be ideal, if there was a mock for Elixir's `File` that performed *read* operations against the Disk (or an overwriting entry in the memory layer if there was a write to that location) but *write* to a memory-based filesystem… similar to what you can use OverlayFS, UnionFS and AUFS for. In the end though, this doesn't exist as far as I know.

The other solutions depend too much on what kind of system you use for development, except for Docker. With Docker though, we would have to run Docker containers within Docker, which works but means some more setup…

**Long story short**, I think you raised a really good point there: "close enough to reality"… To run the tests on different dev machines (OSes, filesystems…) definitely has its merits and maybe we don't even want to go down that other road.

Thank you for sharing your helpful thoughts!
(…and sorry for the lengthy brain dump)

**Comment by klappradla** [comment]
Oh ok, now I get your point 💡 

Well, I'm not familiar with any of the tools you list (except for Docker which I still don't think is a good option here).. A memory based file system could be a good option here though - however, it might come with an extra hassle for dev-setup which is again sort of a downer.

I personally always had good experience with just using temp-files (and directories) - did this last at Liefery if I remember correctly. The way I understood this was: as long as you fully stay in the `tmp` area, you won't really mess with anything and - especially if you have an *always-teardown™* you also won't pollute the local disc to hard 🚮
